src/passes/StackIR.cpp

```diff
@@ -228,6 +235,63 @@ class StackIROptimizer {
     }
   }
 
+  // Get rid of unnecessary push/drop pairs: if something is pushed on the stack
+  // only to be dropped, and it has no side effects, then we don't need it.
+  void dropDrops() {
+    // Track the wasm value stack. We track by index so that we can modify.
+    std::vector<Index> valueStack;
+    for (Index i = 0; i < insts.size(); i++) {
+      auto* inst = insts[i];
+      if (!inst) {
+        continue;
+      }
+      if (inst->op == StackInst::Basic && inst->origin->is<Drop>()) {
+        // The value on the top of the stack is being dropped, do we need it? If
+        // it consumes more than 1 value then we'd need to add more drops to get
+        // rid of it, which might not be worth it.
+        // TODO: the case of 2 seems optimizable, by replacing the old value
+        //       with a drop.
+        assert(!valueStack.empty());
+        auto droppedIndex = valueStack.back();
+        valueStack.pop_back();
+        auto* droppedInst = insts[droppedIndex];
+        // TODO: handle drops of control flow elements like blocks, which show
+        //       up as multiple insts.
+        if (droppedInst && !isControlFlow(droppedInst)) {
+          auto consumedByDropped = getNumConsumedValues(droppedInst);
+          if (consumedByDropped < 2) {
+            EffectAnalyzer effects(passOptions, module->features);
+            effects.visit(droppedInst->origin);
+            if (!effects.hasSideEffects()) {
+              // Great, we don't need to push it in the first place!
+              insts[droppedIndex] = nullptr;
+              // If it dropped one value, keep the drop for that value;
+              // otherwise remove the drop too.
+              if (consumedByDropped == 0) {
+                insts[i] = nullptr;
+              }
+            }
+          }
+        }
+        continue;
+      }
+      auto consumed = getNumConsumedValues(inst);
+      // TODO: currently we run dce before this, but if we didn't, we'd need
+      //       to handle unreachable code here - it's ok to pop multiple values
+      //       there even if the stack is at size 0.
+      while (consumed > 0) {
```

Comment:
This while loop could be a for loop, which would make it easier to understand IMO (not that it's particularly difficult to understand now).